### DISCUSSION TITLE
Custom Search Provider - How to use Verticals Information

### DISCUSSION THREAD

**Discussion by Fringie** [microsoft-search/pnp-modern-search#2600] [Q&A] [answered]
Hello,

I would like to create a search provider that connects to a list, then depending on the Vertical you have selected the search provider will give you relevant selections for example:

I have selected the "HR" vertical, , now the search provider displays HR results:

![image](https://user-images.githubusercontent.com/7203124/204292748-315b5391-fa9f-4b74-9212-152dc72909ff.png)

My question is, how do you access the Vertical value via code?

**Comment by edarroudi** [comment, marked as answer]
Hey @Fringie , you can use DynamicProperties like in SearchResultsWebpart.ts
Have a look at `"private _verticalsConnectionSourceData: DynamicProperty<IDataVerticalSourceData>;"`

**Reply by Fringie** [reply]
Hey dude,

Thanks for posting.

Looks interesting. How do I initialize the value? i.e.
`"let _verticalsConnectionSourceData: DynamicProperty<IDataVerticalSourceData> = this.parent.getVerticals().Verticals.value;"`

A bit of an unrelated question, but how did you know that too? Is there documentation on it or did you just find out by using it etc? When I search for info on the question I can't find anything.

**Reply by edarroudi** [reply]
So what you have to do is to add a custom dynamic property to your custom datasource. You can and should just copy the code shown in the SearchResultsWebpart.ts

here is the code how to create the vertical connection fields: https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/webparts/searchResults/SearchResultsWebPart.ts#L1942

Once you have that dynamic property and connect to the vertical webpart, you can retrieve the actual value: 
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/webparts/searchResults/SearchResultsWebPart.ts#L404-L405

You can have a look at the VerticalWebpart and look at the getPropertyValue fucntion to see which properties the vertical webpart is exposing: https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/webparts/searchVerticals/SearchVerticalsWebPart.ts#L160-L172

In general I just parse the code of the repositiories as there are many tricks / hacks involved in SPFX that are not documentated. Also you'll find many helper methods that make your life easier. Of course you have to look at microsoft learn / docs as well to understand the concepts like [this](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/dynamic-data)

**Reply by Fringie** [reply]
Hi, thank you for the detailed post. I'll spend some time digesting what you've said and will give it a try then I'll respond back here.

**Reply by Fringie** [reply]
> So what you have to do is to add a custom dynamic property to your custom datasource. You can and should just copy the code shown in the SearchResultsWebpart.ts
> 
> here is the code how to create the vertical connection fields:
> 
> https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/webparts/searchResults/SearchResultsWebPart.ts#L1942
> 
> Once you have that dynamic property and connect to the vertical webpart, you can retrieve the actual value:
> 
> https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/webparts/searchResults/SearchResultsWebPart.ts#L404-L405
> 
> You can have a look at the VerticalWebpart and look at the getPropertyValue fucntion to see which properties the vertical webpart is exposing:
> 
> https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/webparts/searchVerticals/SearchVerticalsWebPart.ts#L160-L172
> 
> In general I just parse the code of the repositiories as there are many tricks / hacks involved in SPFX that are not documentated. Also you'll find many helper methods that make your life easier. Of course you have to look at microsoft learn / docs as well to understand the concepts like [this](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/dynamic-data)

I spent a day trying to get this to work with no success. I was missing a lot of the imports because I'm using a library component? I also referred to [this guide](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/dynamic-data) but it's for a webpart so I don't have a `this.context.dynamicDataSourceManager`.

**Reply by edarroudi** [reply]
hey @Fringie,

In the seach provider extension you definitley have the context as the BaseSuggestionProvider provides the webpartcontext: https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-extensibility/src/models/suggestions/BaseSuggestionProvider.ts#L30

The pnp-modern-search package has some helper methods so you can't import them directly. Sometimes it's not necessary.
Here I did a quick boilerplate for you - It's not perfect yet, as I think it depends on when what is called (order). So a lot of room for improvement.

```
import { BaseSuggestionProvider, IDataVertical, ISuggestion } from '@pnp/modern-search-extensibility';
import { IPropertyPaneDropdownOption, IPropertyPaneGroup, PropertyPaneDropdown } from '@microsoft/sp-property-pane';
import { IDynamicDataSource } from '@microsoft/sp-dynamic-data';
import { DynamicProperty } from '@microsoft/sp-component-base';

export interface IXXSuggestionProviderProperties {
  verticalsDataSourceReference: string;
}

export interface IDataVerticalSourceData {
  /**
   * The current selected vertical
   */
  selectedVertical: IDataVertical;

  /**
   * The selected vertical fro the Web Part
   */
  selectedVerticalKeys: string[];
}

export class XXSuggestionProvider extends BaseSuggestionProvider<IXXSuggestionProviderProperties> {
  private _options: IPropertyPaneDropdownOption[];
  protected _verticalsConnectionSourceData: DynamicProperty<IDataVerticalSourceData>;

  public async onInit(): Promise<void> {
    this.serviceScope.whenFinished(() => {
      // ...
    });

    this._options = await this.getAvailableDataSourcesByType('pnpSearchVerticalsWebPart');
    this.ensureDynamicDataSourcesConnection();
  }

  // Copied and modified from https://github.com/microsoft-search/pnp-modern-search/blob/main/search-parts/src/services/dynamicDataService/DynamicDataService.ts
  public async getAvailableDataSourcesByType(propertyId: string): Promise<IPropertyPaneDropdownOption[]> {
    const propertyOptions: IPropertyPaneDropdownOption[] = [];

    if (!this.context.dynamicDataProvider.isDisposed) {
      this.context.dynamicDataProvider.getAvailableSources().forEach(async (sourceInfo) => {
        const source: IDynamicDataSource = this.context.dynamicDataProvider.tryGetSource(sourceInfo.id);
        if (source) {
          const properties = await source.getPropertyDefinitionsAsync();
          properties.forEach((prop) => {
            if (prop.id === propertyId) {
              propertyOptions.push({
                key: `${source.id}:${prop.id}`,
                text: prop.title,
              });
            }
          });
        }
      });
    }

    return propertyOptions;
  }

  /**
   * Make sure the dynamic properties are correctly connected to the corresponding sources according to the proeprty pane settings
   * Author: Copied and modifier code from @pnp-modern-search
   */
  private ensureDynamicDataSourcesConnection() {
    // Verticals Web Part data source
    if (this.properties.verticalsDataSourceReference) {
      if (!this._verticalsConnectionSourceData) {
        this._verticalsConnectionSourceData = new DynamicProperty<IDataVerticalSourceData>(this.context.dynamicDataProvider);
      }

      this._verticalsConnectionSourceData.setReference(this.properties.verticalsDataSourceReference);
    }
  }

  public get isZeroTermSuggestionsEnabled(): boolean {
    return false;
  }

  public async getSuggestions(queryText: string): Promise<ISuggestion[]> {
    if (this._verticalsConnectionSourceData) {
      //&& this.properties.selectedVerticalKeys.length > 0
      const verticalData = this._verticalsConnectionSourceData.tryGetValue();
      console.log('verticalData', verticalData);
    }

    return [
      {
        displayText: 'test',
        groupName: 'test',
      },
    ];
  }

  public getPropertyPaneGroupsConfiguration(): IPropertyPaneGroup[] {
    return [
      {
        groupName: 'XXX Suggestion Settings',
        groupFields: [
          PropertyPaneDropdown('providerProperties.verticalsDataSourceReference', {
            options: this._options,
            label: 'Verticals Data Source Reference',
          }),
        ],
      },
    ];
  }
}
```


For the selected vertical
![image](https://user-images.githubusercontent.com/935493/206426881-a8a7b324-ba99-4bb5-bf31-8a37a4c149f2.png)

when you type in something the result in the console will be


![image](https://user-images.githubusercontent.com/935493/206426726-f7e1a039-93fe-4f67-ba6e-4a0478071ce7.png)

Should be enough to get it on the right path.

**Reply by Fringie** [reply]
Thank you, that worked for me. I was trying to access the context in the root class as I assumed that was where it was set, It didn't click that we are inheriting the context and that the options vary depending on what we are inheriting from.. Thank you for providing this code because it showed me how to do it the correct way; I learned a few things.

**Comment by edarroudi** [comment]
Hey @Fringie,  Great! Can you mark the solution above as answered so this gets closed? Thank you.
